File: deno/2022/day01.ts
import { readInput } from "../utils/read-input.ts";

async function calculateCalories(year: number, day: number): Promise<number[]> {
    return (await readInput(year, day))
        .split("\n\n")
        .map((individual) =>
            individual
                .split("\n")
                .map((cals) => parseInt(cals, 10))
                .reduce((acc, val) => acc + val, 0)
        )
        .sort((a, b) => b - a);
}

export default {
    "1": {
        async partA(year: number, day: number): Promise<string> {
            const [mostCalories] = await calculateCalories(year, day);
            return mostCalories.toString();
        },
        async partB(year: number, day: number): Promise<string> {
            return (await calculateCalories(year, day))
                .slice(0, 3)
                .reduce((acc, val) => acc + val, 0)
                .toString();
        },
    },
};

File: rust/src/solutions/y_2022/day_01.rs
use crate::common::{load, Solution};

pub struct Day01 {}

fn calculate_calories(year: u32, day: u32) -> Vec<u32> {
    let mut data = load(year, day)
        .split("\n\n")
        .map(|individual| -> u32 {
            individual
                .split('\n')
                .map(|x| x.parse::<u32>().unwrap())
                .reduce(|acc, item| acc + item)
                .unwrap()
        })
        .collect::<Vec<u32>>();

    data.sort();
    data.reverse();
    data
}

impl Solution for Day01 {
    fn name(&self) -> String {
        "Calorie Counting".to_owned()
    }

    fn part_a(&self, year: u32, day: u32) -> String {
        calculate_calories(year, day).first().unwrap().to_string()
    }

    fn part_b(&self, year: u32, day: u32) -> String {
        calculate_calories(year, day)
            .drain(0..2)
            .reduce(|acc, calories| acc + calories)
            .unwrap()
            .to_string()
    }
}
